### ISSUE TITLE
Multiple processes sharing PINCache

### ISSUE THREAD

**Issue by ebgraham** [pinterest/PINCache#2]
To support notification center and Apple Watch extensions, we'd like to move our cache location to a shared app group container.  However this enables multiple processes to simultaneously attempt to read and write on the same files.  Does PINCache have any protection against this, by using NSFileCoordinator, for example?

**Comment by garrettmoon** [comment]
No, it uses a runtime lock to protect file access. However, that sounds like a good enhancement to be made!

**Comment by garrettmoon** [comment]
Looked into using NSFileCoordinator a bit.

One problem:
"If your app enters the background with an active file presenter, any other processes that perform a coordinated read or write on the presented file can deadlock. To prevent this situation, call removeFilePresenter: to remove the file presenter in the applicationDidEnterBackground: method or in response to a UIApplicationDidEnterBackgroundNotification notification. Call addFilePresenter: to add the file presenter again in the applicationWillEnterForeground: method or in response to a UIApplicationWillEnterForegroundNotification notification."

This may be an issue with extensions as notifications for backgrounding don't occur if I'm not mistaken.

**Comment by ebgraham** [comment]
Extensions do actually get little-known lifecycle notifications:
```
NSExtensionHostDidBecomeActiveNotification
NSExtensionHostDidEnterBackgroundNotification
NSExtensionHostWillEnterForegroundNotification
NSExtensionHostWillResignActiveNotification
```

**Comment by garrettmoon** [comment]
Ahh! Nice!

The next concern I have is around performance. That's the next question to tackle :)

**Comment by coreylowegp** [comment]
Has this been implemented?
